### ISSUE TITLE
DebugLock means that Mutex isn't a mutex at all when NDEBUG is defined

### ISSUE THREAD

**Issue by karlhendrikse** [AcademySoftwareFoundation/OpenColorIO#1235]
This is the code of `DebugLock`:
```
// In debug mode, try to trap recursive cases and lock/unlock debalancing cases.
class DebugLock
{
public:
    DebugLock() = default;
    DebugLock(const DebugLock &) = delete;
    DebugLock& operator=(const DebugLock &) = delete;
    ~DebugLock() { assert(!m_locked); }

    void lock()     { assert(!m_locked); m_mutex.lock();   m_locked = true;  }
    void unlock()   { assert(m_locked);  m_mutex.unlock(); m_locked = false; }
    void try_lock() { assert(!m_locked); m_locked = m_mutex.try_lock(); }

    bool locked() const { return m_locked; }

private:
    // An exclusive and non-recursive ownership lock.
    std::mutex m_mutex;
    bool       m_locked = false;
};
```
Note that if two threads try to lock the mutex at the same time, one of them will assert. That defeats the entire purpose of a mutex. I assume the code is trying to prevent the same thread from recursively locking the mutex multiple times, but for that `m_locked` would need to be thread-local.

**Comment by DarkDefender** [comment]
I just want to chime in and say that I ran into this issue when running Blender.
It would trigger this assert when trying to render certain scenes.

**Comment by doug-walker** [comment]
This was apparently fixed by PR #1407.
